posts/coding-intentionally-in-bash-grains.md

```diff
@@ -0,0 +1,127 @@
+ >
+ > There are 64 squares on a chessboard.
+ >
+ > Write code that shows:
+ >
+ > - How many grains were on each square, and
+ > - The total number of grains
+
+The student came up with a pretty clever way to calculate the total.
+
+```bash
+bc <<< 'ibase=16;FFFFFFFFFFFFFFFF'
+```
+
+*For those of you that have completed this exercise on the Bash track and know tricky other ways to not use `bc` (the command line calculator), I'm not going to show that here.  We'll stick with `bc` for these examples, for clarity.
+
+This solution works because the whole exercise revolves around powers of two — and where there are powers of two, there's binary, and where there is binary, there's hexadecimal!
+
+The first square has 1 (`0b1`) grains on it.  Running total: 1.  
+
+The second square has 2 (`0b10`) grains on it.  Running total: 3 (`0b11`).  
+
+The third square has 4 (`0b100`) grains on it.  Running total: 7 (`0b111`).  
+
+See how it works?  With every square, we're slowly but surely filling in the total with binary ones in every place.  The total should be equivalent to 64 ones, or, as this student showed, 16 `F`'s in hexadecimal (4 binary digits per hex digit).
+
+## Design Intent
+
+I'm a firm believer that **Design Intent** is one of the most important things to consider when implementing an engineering design.  This is true across disciplines.  *How* you go about implementing something doesn't mean much to the end user, but it should speak *volumes* to every engineer that touches your design at any point.  There are often a lot of ways to achieve the same functionality, and it might seem like any of the options would be sufficient to get the job done.  However, I believe that every single decision you make should have a reason (even if it's a small decision with a small reason), and that reason should communicate an objective or requirement.
+
+I see this a lot when doing mechanical design.  When I'm working on a 3D CAD design of a hole for a .250" pin to fit into, if I wanted a close fit, I could define it by saying that the hole should be .2504" in diameter.  That would be the correct dimension, but it doesn't communicate design intent.  If, instead, I defined the hole by saying that it should be .0004" bigger than the pin diameter, the design intent is more clearly shown.  The hole is made to fit the pin.  If we change designs and decide to go with a .750" pin, the hole will be ready to be updated easily, since the relationship between them and the **intent** that they should fit together is clearly communicated.
+
+In the case of the student's *Grains* solution, what the code is telling me is that hexadecimal is important here, or that the solution fundamentally revolves around 16.  Neither of those are the case, and we brainstormed ideas for how to more clearly communicate intent.  Here are a few options:
+
+### First Option: Binary
```

Comment:
Style guide thing: we use sentence case for all titles on the website. @iHiD should we standardize to that on the blog, too?